runbld fails when scm clone is true but wipe-workspace is false

Basically, it cannot clone when the workspace already exists and is populated.  It should probably run `git pull` instead.

<details>

```
[elastic+elasticsearch+5.4+packaging-tests] $ /usr/local/bin/runbld --last-good-commit elastic+elasticsearch+5.4+multijob-intake /tmp/jenkins846470968449712863.sh
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
runbld>>> runbld started
runbld>>> 1.3.7/cd451cb20de4e95b3a5b26f6f2bf7812b27d466d
runbld>>> BUILD: https://5086a1f436ee16623a447bdf25881bbc.us-east-1.aws.found.io:9243/build-1463180427010/t/20170606193430-2367880D
runbld>>> cloning git@github.com:elastic/elasticsearch.git
Exception in thread "main" java.lang.AssertionError: Assert failed: ["clone" ("git@github.com:elastic/elasticsearch.git" "<https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.4+packaging-tests/ws/"> "--reference" "/var/lib/jenkins/.git-references/elasticsearch.git" "--branch" "5.3")]:
out:
err: fatal: destination path '<https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.4+packaging-tests/ws/'> already exists and is not an empty directory.

(= 0 (:exit r))
	at clj_git.core$run$err__2817.invoke(core.clj:36)
	at clj_git.core$run.invokeStatic(core.clj:43)
	at clj_git.core$run.invoke(core.clj:29)
	at clj_git.core$run.invokeStatic(core.clj:31)
	at clj_git.core$run.invoke(core.clj:29)
	at clj_git.core$git_clone.invokeStatic(core.clj:225)
	at clj_git.core$git_clone.invoke(core.clj:208)
	at runbld.main$fn__11744$bootstrap_workspace__11749$fn__11750.invoke(main.clj:84)
	at runbld.main$fn__11744$bootstrap_workspace__11749.invoke(main.clj:63)
	at runbld.main$fn__11908$fn__11909.invoke(main.clj:138)
	at runbld.main$fn__11899$fn__11900.invoke(main.clj:137)
	at runbld.main$fn__11890$fn__11891.invoke(main.clj:136)
	at runbld.main$fn__11881$fn__11882.invoke(main.clj:135)
	at runbld.main$fn__11872$fn__11873.invoke(main.clj:134)
	at runbld.main$fn__11863$fn__11864.invoke(main.clj:133)
	at runbld.main$_main.invokeStatic(main.clj:160)
	at runbld.main$_main.doInvoke(main.clj:150)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at runbld.main.main(Unknown Source)
Build step 'Execute shell' marked build as failure
```

</details>

--- a/src/clj/clj_git/core.clj
+++ b/src/clj/clj_git/core.clj
@@ -234,3 +234,20 @@
    (git-pull repo []))
   ([repo args]
    (git repo "pull" args)))
+
+(defn git-fetch
+  "Runs `git fetch' on the provided repo.  `args' is a vector of
+  options as they would be specified on the command line.
+
+  clj: (git-fetch repo [\"--all\"])
+  cli: git fetch --all"
+  ([repo]
+   (git-fetch repo []))
+  ([repo args]
+   (git repo "fetch" args)))
+
+(defn git-remote
+  "Runs `git remote' on the provided repo.  `args' is a vector of
+  options as they would be specified on the command line."
+  [repo args]
+  (git repo "remote" args))

--- a/src/clj/runbld/main.clj
+++ b/src/clj/runbld/main.clj
@@ -60,6 +60,22 @@
       (io/rmdir-contents workspace)))
   opts)
 
+(defn update-workspace
+  "Updates an existing workspace and gets it onto the correct
+  branch."
+  [local branch depth]
+  (let [repo (git/load-repo local)]
+    (io/log "repo already cloned" local)
+    (io/log "updating")
+    (git/git-remote repo ["set-branches" "origin" branch])
+    (let [fetch-args (concat (when depth
+                               ["--depth" (str depth)])
+                             ["origin" (str branch)])]
+      (git/git-fetch repo fetch-args))
+    (git/git-checkout repo branch)
+    (git/git-pull repo)
+    (io/log "done updating")))
+
 (s/defn bootstrap-workspace
   [opts :- {:process OptsProcess
             :build Build
@@ -73,16 +89,18 @@
                    (-> opts :build :branch))
         depth (-> opts :scm :depth)]
     (when clone?
-      (let [clone-args (->> [(when (and reference
-                                        (.exists (jio/as-file reference)))
-                               ["--reference" reference])
-                             (when branch ["--branch" (str branch)])
-                             (when depth ["--depth" (str depth)])]
-                            (filter identity)
-                            (apply concat))]
-        (io/log "cloning" remote)
-        (git/git-clone local remote clone-args)
-        (io/log "done cloning"))))
+      (if (.exists (jio/file local ".git"))
+        (update-workspace local branch depth)
+        (let [clone-args (->> [(when (and reference
+                                          (.exists (jio/as-file reference)))
+                                 ["--reference" reference])
+                               (when branch ["--branch" (str branch)])
+                               (when depth ["--depth" (str depth)])]
+                              (filter identity)
+                              (apply concat))]
+          (io/log "cloning" remote)
+          (git/git-clone local remote clone-args)
+          (io/log "done cloning")))))
   opts)
 
 (s/defn log-script-execution

--- a/test/config/scm.yml
+++ b/test/config/scm.yml
@@ -28,6 +28,16 @@ profiles:
         # Git should ignore reference repos that aren't accessible.
         reference-repo: /some/fake/path
 
+  - '^elastic\+foo\+5\.x$':
+      scm:
+        clone: true
+        url: https://github.com/elastic/elasticsearch.git
+        branch: 5.x
+        depth: 1
+        wipe-workspace: true
+        # Git should ignore reference repos that aren't accessible.
+        reference-repo: /some/fake/path
+
   - '^owner\+project\+.*':
       scm:
         clone: true

--- a/test/runbld/test/main_test.clj
+++ b/test/runbld/test/main_test.clj
@@ -312,6 +312,22 @@
                 (is (.startsWith
                      (let [[_ _ _ subj _ _] @email] subj) "FAILURE"))
                 (is (.contains (slack-msg) "FAILURE")))
+              (testing "Running again w/o wipe-workspace should update the repo"
+                (let [[opts2 res2] (run
+                                     (conj
+                                      ["-c" "test/config/scm.yml"
+                                       "-j" "elastic+foo+5.x"
+                                       "-d" workspace]
+                                      (if (opts/windows?)
+                                        "test/fail.bat"
+                                        "test/fail.bash")))
+                      branch (-> opts2 :scm :branch)
+                      depth (-> opts2 :scm :depth)
+                      repo2 (load-repo workspace)]
+                  (is (not (= (-> opts :scm :branch)
+                              (-> opts2 :scm :branch))))
+                  (is (= branch (git-branch repo2)))
+                  (is (= depth (count (git-log repo2))))))
               (testing "wiping the workspace"
                 (wipe-workspace-orig opts)
                 ;; Should only have the top-level workspace dir left

--- a/test/runbld/test/main_test.clj
+++ b/test/runbld/test/main_test.clj
@@ -222,28 +222,31 @@
 
 (s/deftest set-scm-build
   (testing "the scm build can be set multiple ways:"
-    (let [args (atom [])]
-      (with-redefs [git-clone (fn [_ _ clone-args]
-                                (reset! args clone-args))
+    (let [branch (atom nil)]
+      (with-redefs [git-clone (fn [local remote clone-args]
+                                (reset! branch
+                                        (get (apply hash-map clone-args)
+                                             "--branch")))
+                    main/update-workspace (fn [local b depth]
+                                            (reset! branch b))
                     main/wipe-workspace (fn [workspace] nil)
                     scheduler/as-map identity]
         (testing "supplying branch in yml"
-          (git/with-tmp-repo [d "tmp/git/owner+project+branch"]
-            (let [raw-opts (-> (opts/parse-args
-                                ["-c" "test/config/scm.yml"
-                                 "-j" "owner+project+master"
-                                 "-d" "tmp/git/owner+project+branch"
-                                 "test/fail.bash"])
-                               (assoc :logger runbld.io/log)
-                               ;; make schema happy
-                               (assoc :scheduler (default-sched/make {})))
-                  opts (build/add-build-meta raw-opts)]
-              (main/bootstrap-workspace
-               (assoc-in opts [:scm :wipe-workspace] false))
-              (is (= "master"
-                     (-> opts :scm :branch)
-                     (-> opts :build :branch)
-                     (get (apply hash-map @args) "--branch"))))))
+          (let [raw-opts (-> (opts/parse-args
+                              ["-c" "test/config/scm.yml"
+                               "-j" "owner+project+master"
+                               "-d" "tmp/git/owner+project+branch"
+                               "test/fail.bash"])
+                             (assoc :logger runbld.io/log)
+                             ;; make schema happy
+                             (assoc :scheduler (default-sched/make {})))
+                opts (build/add-build-meta raw-opts)]
+            (main/bootstrap-workspace
+             (assoc-in opts [:scm :wipe-workspace] false))
+            (is (= "master"
+                   (-> opts :scm :branch)
+                   (-> opts :build :branch)
+                   @branch))))
         (testing "extracting branch from job"
           (git/with-tmp-repo [d "tmp/git/owner+project+branch"]
             (let [raw-opts (-> (conj
@@ -262,7 +265,7 @@
               (is (nil? (-> opts :scm :branch)))
               (is (= "branch"
                      (-> opts :build :branch)
-                     (get (apply hash-map @args) "--branch"))))))))))
+                     @branch)))))))))
 
 (s/deftest ^:integration execution-with-scm
   (testing "real execution all the way through with cloning via scm config"

